File: daily/Linked List/25-Reverse-nodes-in-k-group.go
package daily

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}

	// count length
	length := 0
	curr := head
	for curr != nil {
		length++
		curr = curr.Next
	}

	dummy := &ListNode{Val: 0, Next: head}
	groupPrev := dummy
	curr = head

	for i := 0; i < length/k; i++ {
		groupTail := curr  // curr is the future tail of this reversed group
		var prev *ListNode // nil in Go, equivalent to Python's None

		for j := 0; j < k; j++ {
			nxt := curr.Next
			curr.Next = prev
			prev = curr
			curr = nxt
		}

		groupPrev.Next = prev // stitch left side
		groupTail.Next = curr // stitch right side
		groupPrev = groupTail // slide boundary forward
	}

	return dummy.Next
}

File: daily/Linked List/25-Reverse-nodes-in-k-group.py
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return None
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next
        
        groups = length // k
        dummy = ListNode(0, head)
        group_prev = dummy
        curr = head
        for _ in range(groups):
            group_detail = curr
            prev = None
            for _ in range(k):
                if curr:
                    nxt = curr.next
                    curr.next = prev
                    prev = curr
                    curr = nxt
            group_prev.next = prev
            group_detail.next = curr
            group_prev = group_detail
        return dummy.next
            